### PR TITLE
[Feature] Add ability to configure comment style

### PR DESCRIPTION
**Pull request by STBoyden**
Changes:

- Adds the config option `use_block_header` to optionally prefer the block-style comment in a given language (if `true`) over the single-line comment (if `false`).

	By default, the value is `true` to maintain the original behaviour, which preferred block-style comments over multiple single-line comments.

- Adds a test to check that the above is actually working (with an edge case for languages that only have one type of comment i.e. Python).

### CODE REVIEW

**Comment by attilarepka** [comment]
Hi @STBoyden!
Looks good, however —moving the comment style selection into `comment_headers()` would simplify things a lot.

Languages would just declare their available comment styles, and the logic to pick between block or line style would be handled centrally.
That means less repetition, easier maintenance.

What do you think?

**Comment by STBoyden** [comment]
> Hi @STBoyden! Looks good, however —moving the comment style selection into `comment_headers()` would simplify things a lot.
> 
> Languages would just declare their available comment styles, and the logic to pick between block or line style would be handled centrally. That means less repetition, easier maintenance.
> 
> What do you think?

That sounds like a logical option. Probably should have thought of that in the first place lmao.

Would something like this be appropriate?:

Remove all `use_block_header` references in `languages.lua` and add a `comment_single` field:
```lua
-- for example in languages.cpp
local comments = {
	comment_start = "/*",
	comment = "*",
	comment_single = "//"
	comment_end = "*/",
}
```

And make the special-case languages like Python and R something like this:
```lua
-- for example in languages.python/languages.r
local single = "#"
local comments = {
	comment_start = nil,
	comment = single,
	comment_single = single,
	comment_end = nil,
}
```

Then in `comment_headers`, like you suggested, do the lookup depending on the config value:
```lua
local result = {}

if header.config.use_header_block and comments.comment_start ~= nil then
    table.insert(result, comments.comment_start)
end

for _, entry in ipairs(header_lines) do
	local prefix = header.config.use_header_block and comments.comment or comments.comment_single

    if entry == "" then
        table.insert(result, prefix)
    else
        table.insert(result, prefix .. " " .. entry)
    end
end

if header.config.use_header_block and comments.comment_end ~= nil then
    table.insert(result, comments.comment_end)
end

table.insert(result, "")

return result
```

**Comment by STBoyden** [comment]
Okay, sounds good. I'll implement this in a few hours approx - Currently travelling. 😃

**Comment by STBoyden** [comment]
> > Okay, sounds good 👍🏻. I'll implement this in a few hours approx - currently travelling
> 
> Thanks for the update! So I have tested the current version of the PR, and when `use_block_header` is set to false, or for languages without block comments (like Python), inserting or updating header clears the whole file contents.

Yeah currently working on a fix for that as we speak, I noticed it last night.

**Comment by attilarepka** [comment]
As a final note, I encountered additional issues during testing. For example, if a file already contains a block header and the config is changed to `use_block_header = false`, updating the header results in duplication (see screenshot below):
image

The same happens when adding license headers under these conditions:
image

**Comment by STBoyden** [comment]
:+1: I'll get on this

File: lua/header.lua [hunk 1]
```diff
@@ -77,56 +87,67 @@ local function escape_special_characters(pattern)
...
+--             end
+--         else
+--             if line:find(end_pat) then
+--                 return i -- End of block header (inclusive)
+--             end
+--         end
+--     end
+--
+--     return 0 -- No complete block comment found
+-- end
```
Comment (attilarepka) [inline lua/header.lua:113]: Can be removed.

File: lua/header.lua [hunk 2]
```diff
@@ -77,56 +87,67 @@ local function escape_special_characters(pattern)
...
+            or (block_end_pat   ~= "" and line:match("^%s*" .. block_end_pat))
+        then
+            last_comment_line = i
+        elseif line:match("^%s*$") then -- shortcircuit on the first white-space only/empty line
             last_comment_line = i
+            break
         else
             break
         end
+        -- stylua: ignore end
```
Comment (attilarepka) [inline lua/header.lua:143]: Let's have consistent `stylua` formatting, do not use ignore's.

Comment (STBoyden) [reply lua/header.lua:143]: Sure - sorry it was to align the conditions (I can be a bit OCD at times with that).

**Comment by attilarepka** [comment]
Thanks!
